[Core] ABC table and ABC update

otocephaly opened the pull request:
initial development of the table

## Pull request type

- [x] This is new functionality or an addition to existing functionality

## Pull request details

<!-- If this PR is for new functionality or a bugfix, please complete the section below.  For a patch bump, delete this comment and the section below.  The context for your change is important to help the reviewer understand what the change is supposed to do - please provide an appropriate link or description to help the review process. -->
- [x] This is in response to a discussion or thread on Discord: *[Link to thread or start of discussion]*(https://discord.com/channels/441414116914233364/1285154582292332554/1285154582292332554)
- [x] The goal of this PR is detailed below:
    - consolidate ABC violations (weaving, interrupts, do nothing from BLM)

## Testing / Validation

<!-- If this PR contains any new functionality or bugfixes, please include log links that reviewers can use to help verify your changes.  Reviewers may also find additional logs to check your PR with, but having initial logs is important to speed the review process, especially for corner cases or hard to trigger analysis flags.  If this PR is a patch bump with no changes or potency only changes, you can delete this section. -->
- [x] I used the log(s) listed below to develop and test this bugfix:
    - Healers
        - (AST - M2S) https://www.fflogs.com/reports/a:nXbkaRJt4fyVBGjH?fight=1&source=7
        - (SGE - FRU) https://www.fflogs.com/reports/j7cML4xmpygqGDFa?fight=17&source=13
        - (SCH - M4S) https://www.fflogs.com/reports/23kgHBQGypPNm4rv?fight=45&source=861
        - (WHM - M2S) https://www.fflogs.com/reports/21qaj4H9pXhTgGrJ?fight=4&source=28
    - Phys Ranged
        - (BRD - M4S) https://www.fflogs.com/reports/1K2rpnW9Jcx4FzQC?fight=22&source=33
        - (DNC - M1S) https://www.fflogs.com/reports/bz1QZAJ64pcPKqVj?fight=1&source=2
        - (MCH - EX2) https://www.fflogs.com/reports/qdc1JHLAmak6R79w?fight=34&source=858
    - Magic Ranged
        - (BLM - M3S) https://www.fflogs.com/reports/Mg7Pj2p1tzAGWBdH?fight=63&source=495
        - (PCT - EX1) https://www.fflogs.com/reports/AXgym4h3YqDwGVFx?fight=3&source=8
        - (RDM - M3S) https://www.fflogs.com/reports/qJgYWRBVT6C2yAfk?fight=7&source=5
        - (SMN - FRU) https://www.fflogs.com/reports/xkYgjVFymCGZaw1R?fight=47&source=1807
    - Tanks
        - (DRK - FRU) https://www.fflogs.com/reports/a:2xAzgYFKw8yhMVpm?fight=4&source=2
        - (GNB - EX2) https://www.fflogs.com/reports/yZxpbA7PhXQ2JFDw?fight=4&source=4
        - (PLD - M1S) https://www.fflogs.com/reports/fKqy729tkMpH4QFT?fight=1&source=2
        - (WAR - COD) https://www.fflogs.com/reports/th7PrLHApRMW4GXk?fight=1&source=2
    - Melee
        - (DRG - EX3) https://www.fflogs.com/reports/VJy1TGg94QL7jqPx?fight=43&source=304
        - (MNK - M3S) https://www.fflogs.com/reports/a:6nFwzD7qJdjTmvP8?fight=8&source=19
        - (NIN - M2S) https://www.fflogs.com/reports/7n2tcA3FYhJjCRaP?fight=3&source=73
        - (RPR - M4S) https://www.fflogs.com/reports/PaWV3nL4vdhbQKjm?fight=16&source=116
        - (SAM - M1S) https://www.fflogs.com/reports/a:QBYmpWHcKLjF6qRC?fight=12&source=130
        - (VPR - FRU) https://www.fflogs.com/reports/V4GX7Y9KLqhxdCZP?fight=3&source=5
    - Blue Mage
        - (BLU1 - E9S) https://www.fflogs.com/reports/XkA7qrxVHLpzgQ4v?fight=15&source=5
        - (BLU2 - E9S) https://www.fflogs.com/reports/XkA7qrxVHLpzgQ4v?fight=15&source=3
        - (BLU3 - E12S) https://www.fflogs.com/reports/GBk2F1ARdwW4cZYh?fight=21&source=9

## Job Maintenance
- [x] I have collaborated with one or more job maintainers for the job(s) in this PR while developing it

## Additional Notes
- Deaths have been specifically excluded in this new test vs current prod which causes a difference in percentage calculation. See below for the example showing why that difference exists. The idea is that the time spent dead is not active and not included in the fight in test whereas in prod the time spent dead is active and included in the fight. (This is easy to change if we want to consider it active for both purposes.)
- Another key difference between PROD ABC and this TEST ABC, includes:
    - PROD ABC counts all active time as executed estimated GCDs less cut off from beginning of fight/end of fight
    - TEST ABC doesn't count time close to downtime unless greater than a reaction time as time spent not casting
- Therefore, based on these two notes, to get the difference between the two percentages, you can do the following.
    - ((PROD ABC % * fight time excluding downtime) - time dead - amounts attributable to GCD JITTER[1] + time between ending downtime and first GCD + time between last GCD and starting downtime) / fight excluding downtime and time dead
    - [1] these are amounts where the actual GCD and estimated GCD is below the intended GCD JITTER
- There is also an allowance for caster tax (100ms or 0.1s) and GCD JITTER (50ms or 0.05s) which will filter out some of the noise

## Preview
Example log (weaving, death, interrupts, other)
https://www.fflogs.com/reports/a:nXbkaRJt4fyVBGjH?fight=1&source=7
![image](https://github.com/user-attachments/assets/6e65a43e-4a2b-409a-9580-79c41b1ca72e)
![image](https://github.com/user-attachments/assets/e8a8eaa1-3c12-4565-b637-3cf91acbe950)
![image](https://github.com/user-attachments/assets/b3b139e0-af2d-4e84-813d-bb3260a60464)
![image](https://github.com/user-attachments/assets/94521926-5a61-4e36-b0a5-a21d1fae89ff)
![image](https://github.com/user-attachments/assets/9791897f-be10-4dd8-8158-464e624a0f38)

## Example showing difference in ABC percent due to death
FFLog: https://www.fflogs.com/reports/a:nXbkaRJt4fyVBGjH?fight=1&source=7
[A] Fight time = 568 seconds
[B] Downtime = 4 seconds
[C] Time spend dead = 36 seconds
[D] % per prod = 83.78%
[E] % per test = 82.66%

approx active time per prod = ([A] - [B]) * [D] = 473 seconds
approx active time per test = ([A] - [B] - [C]) * [E] = 436 seconds
approx difference = 37 seconds which is approximately the time spent dead

Per prod:
![image](https://github.com/user-attachments/assets/1a04b60a-fdd9-483f-a668-1f844c0240af)

Per test:
![image](https://github.com/user-attachments/assets/03d81f83-a5ca-4b65-932b-a4ec37bbb046)

## Testing details (no screenshots)
Note: this is in order of the testing logs above for ease of reference
note: P: = prod, T: = test where applicable

Job | Uptime % | # weaves | # interrupts | notes
-- | -- | -- | -- | --
Healers
AST | P: 83.78%</br>T: 82.66% | P: 1</br>T: 2 | matched | difference noted in weaves due to prod weave</br>module not checking weaves around ToD or EoF
SGE | P: 96.03%</br>T: 96.60% | P:1</br>T: 10 | matched | there are more weaving issues because previous</br>module allowed for 1 between casts</br>but here we can see that it caused some drifting issues.
SCH | P: 95.89%</br>T: 95.90% | P: 2</br>T: 6 | matched | some weaving issues that look reasonable</br>unfortunately since the module doesn't discriminate</br>these weaving issues are expected
WHM | P: 97.25%</br>T: 96.88% | P: 0</br>T: 5 | matched | difference in weaves because WHM should use glare IV</br>instead of glare III to weave if needed
Phys Ranged
BRD | P: 101.07%</br>T: 99.77% | matched | N/A | difference noted in ABC % likely due to prod not</br>displaying negatives otherwise I see the same issues</br>in the timeline in prod so likely a bug in prod
DNC | P: 99.34%</br>T: 99.57% | matched | N/A | likely due to gcd_jitter difference
MCH | P: 71.63%</br>T: 65.37% | P: 29</br>T: 37 | N/A | this one revealed bug with transcendence drop off
M Ranged
BLM | P: 89.33%</br>T: 89.95% | P: 6</br>T: 11 | matched | difference due to death time</br>note: times you literally did nothing was much</br>less punishing than this module
PCT | P: 95.05%</br>T: 95.07% | P: 2</br>T: 5 | P: 7</br>T: 6 | interrupt is different because for some reason module doesn't pick up last one</br>(but also it's the last second so player shouldn't be punished anyway)
RDM | P: 88.09%</br>T: 88.12% | P: 2</br>T: 7 | matched | since the issues were infrequent</br>these weaving issues seem reasonable
SMN | P: 95.26%</br>T: 93.71% | P: 0</br>T: 2 | matched | weaves seem reasonable
Tanks
DRK | P: 97.31%</br>T: 98.40% | P: 7</br>T: 5 | N/A | difference in weaves due to allowance (GCD_JITTER)</br>as difference noted for two weaves is less</br>than 100 ms (0.1 seconds)
GNB | P: 96.01%</br>T: 96.03% | matched | N/A | mostly same output
PLD | P: 99.27%</br>T: 99.49% | matched | matched | nothing too out of the ordinary here
Melee
WAR | P: 97.18%</br>T: 97.06% | P: 2</br>T: 1 | N/A | one less weave because new ABC has additional allowance</br>close to end of fight</br>not that they needed the extra heal but still
DRG | P: 97.54%</br>T: 97.29% | N/A | N/A | slight difference in uptime % (GCD_JITTER, close to downtime)
MNK | P: 97.08%</br>T: 97.37% | P: 1</br>T: 12 | N/A | updated MNK to exclude instances where oGCDs</br>slightly cut off when GCD below 2s and</br>there are only 2 GCDs
NIN | P: 96.78%</br>T: 96.34% | P: 0</br>T: 4 | N/A | weaving seems a lil punishing, but not horrible
RPR | P: 97.81%</br>T: 98.19% | P: 1</br>T: 6 | P: 0</br>T: 3 | 1.5s GCD seems to throw weaving off a bit</br>RPR didn't have interrupt before because technically melee I guess
SAM | P: 98.98%</br>T: 98.37% | P: 1</br>T: 2 | N/A | difference in weaving likely due to close to the wire</br>no other issues noted</br>shows limit break working as intended
VPR | P: 95.22%</br>T: 96.38% | matched | n/a | difference likely attributable to number of downtimes noted
Blue Mage
BLU1 | P: 91.50%</br>T: 92.01% | P: 19</br>T: 1 | matched | improvement in weaving because BLU is allowed to clip most GCDs</br>also one of the double weaves is on the cusp of jitter
BLU2 | P: 86.40%</br>T: 86.94% | P: 17</br> T: 1 | matched | same as above comment
BLU3 | P: 62.77%</br>T: 61.58% | P: 4</br> T: 1 | N/A | one weave went to other oGCD issues but rest same comment as above

otocephaly committed:
[Core] ABC table initial

initial development of the table

AkairyuGestalter commented on src/data/ACTIONS/root/PCT.ts:
```diff
@@ -88,6 +88,7 @@ export const PCT = ensureActions({
 		onGcd: true,
 		speedAttribute: Attribute.SPELL_SPEED,
 		castTime: 2300,
+		gcdRecast: 3300,
```
I probably should've checked both draft PRs, but the note I left on the other one is still applicable. gcdRecast on an action with a cooldown causes it to get treated like Phlegma. Downstream code should treat cooldown as gcdRecast if gcdRecast is missing and onGcd is true. These changes to PCT's actions should be removed.

otocephaly replied:
reverted and updated logic in the new section

AkairyuGestalter commented on src/parser/core/modules/AlwaysBeCastingTable.tsx:
```diff
@@ -0,0 +1,304 @@
+}
+
+export class ABCTable extends Analyser {
+	static override handle = 'alwaysbecastingtable'
```
Since this is output related to AlwaysBeCasting, it probably should just be in AlwaysBeCasting, rather than a separate Analyser

otocephaly replied:
sounds good! I'll look at combining the two into one Analyser (maybe after this one is done so I don't create any funny business in abc)

AkairyuGestalter commented on src/parser/core/modules/AlwaysBeCastingTable.tsx:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	override output() {
+		if (this.noCastWindows.history.length === 0) { return <><Trans id="core.always-be-casting-table.no-violations">Nothing to note.</Trans> 😉</> }
```
I'd probably just return null/undefined if there aren't any issues so the table doesn't appear in the output at all. That's more to the pattern we follow for other modules where no errors need displaying.

otocephaly replied:
updated!

AkairyuGestalter commented on src/parser/core/modules/AlwaysBeCastingTable.tsx:
```diff
@@ -0,0 +1,304 @@
+					<Table.HeaderCell><Trans id="core.always-be-casting-table.action-header">Relevant action(s)</Trans></Table.HeaderCell>
+					<Table.HeaderCell><Trans id="core.always-be-casting-table.weaving-header"># of Weaves</Trans></Table.HeaderCell>
+					<Table.HeaderCell><Trans id="core.always-be-casting-table.interrupted-header">Interrupted Actions</Trans>*</Table.HeaderCell>
+					<Table.HeaderCell>☠️</Table.HeaderCell>
```
Rather than an emoji, I feel like using the raise action icon would communicate this better, since we already use that for the "Don't die" suggestions.

otocephaly replied:
updated!

AkairyuGestalter commented on src/parser/core/modules/AlwaysBeCastingTable.tsx:
```diff
@@ -0,0 +1,304 @@
+							<br/>
+							({this.parser.formatDuration((notCasting.stop ?? notCasting.start) - notCasting.start)})</Table.Cell>
+						<Table.Cell>
+							<Rotation events={[
```
This is producing notably different results on a log I'm looking at to the main Weaving module: http://localhost:3000/fflogs/CJnbwdNzWmp9aT7K/58/90
- Appears to be showing a single weave at ~2 minutes, which weaving isn't complaining about. This might actually just be because they clipped their GCD to keep Starry perfectly on CD, but it's not clear from this output that's the case
- Appears to be complaining about a "triple weave" at ~8 minutes with only two weaved actions.
- Didn't catch the weave under Hyperphantasia at ~10 minutes

If this column is meant to be a replacement for the separate Weaving Issues table, then we're going to need to take a far closer look at how it's calculating those weaves and determining whether they're legitimate.

otocephaly replied:
100% definitely a lot to iron out
This module needs a bunch of love still and I want to incorporate your other feedback in the discord
The next draft should fix these issues (I have to incorporate your above feedback too and then I'll post the draft in discord again for feedback)
Thank you for this! I appreciate the feedback relating to pct because it's a lil new to me still

otocephaly replied:
So after looking at this, the reason the one around ~10 mins shows in weaving, but not in this is because there's an allowance of 0.05s for GCD_JITTER and they only violated gcd by 0.01s
If I lower the GCD_JITTER to 0, it shows, but then it's very punishing and a lot comes up under other issues
What do you think would be a good allowance if any?

(Additionally, I just noticed ignored actions for PCT and need to update for that, but that'll come in the next time I look at this)

AkairyuGestalter reviewed (changes requested):
Didn't realize until I started looking at this PR that it and the other draft are pretty closely related. Some common concerns, and also some commentary specific to this one.

otocephaly committed:
version 2.0 table

updated based on comments and bugs

otocephaly committed:
lint

lint

otocephaly committed:
updating after prop-types changes

yeah it broken still, but hopefully lint/test will tell me something

otocephaly committed:
finally fix

otocephaly committed:
update for downtime and abc percent

otocephaly committed:
updating modules to adapt to new ABC

why

otocephaly committed:
goodbye weaving

otocephaly committed:
this time ghost are actually gone

otocephaly committed:
extract

extract

otocephaly committed:
I guess the original goal was to get rid of this oop

otocephaly commented:
Hi @AkairyuGestalter 
So I got around to this finally, but the module is seemingly very punishing in a sense
(e.g. check the BLM log and compare other OGCD issues and times you literally did nothing)
I don't think the new ABC is necessarily wrong, but yeah it'd be a rough sell to people since people might see more than like 10 as a big issue
Any who, I still have a bit of clean up to do and some testing to ensure that what is happening is what should be before I give you a final pass (if you're going to accept this lol I understand it's a lot in one PR...)
If you have some capacity, can you do another prelim review like you did before?

otocephaly committed:
clean up including adding downtime

otocephaly committed:
should consider downtime impacts probably

otocephaly committed:
Update changelog.tsx

otocephaly committed:
speeling

otocephaly committed:
blu clean up

otocephaly committed:
adding first oGCD clipping for BLU

otocephaly committed:
diamondback already considered downtime

otocephaly committed:
fixed debugging and icon when die quick downtime

otocephaly committed:
updated for transcendence drop off since it could happen to you

otocephaly committed:
interrupt at end of fight

otocephaly committed:
mnk needs more forgiveness

otocephaly committed:
more random updates based on review

otocephaly commented:
Hi @AkairyuGestalter 
It me again, I think this is a good version now barring review notes and any changes for patch 7.2 (looking at BLM specifically tbh)
I changed this from draft to ready for review and added a testing section to show how the different jobs would look comparatively in the new version of ABC
(No rush and no obligation of course; I appreciate the comments you've given so far :3

otocephaly committed:
Merge branch 'xivanalysis:dawntrail' into core_abc_table

AkairyuGestalter committed:
Merge branch 'dawntrail' of https://github.com/xivanalysis/xivanalysis into pr/otocephaly/2114

AkairyuGestalter committed:
fix some merge stuff

AkairyuGestalter committed:
couple initial readability tweaks for my aging eyes

AkairyuGestalter commented:
After further discussion on Discord since I tried picking review for this back up, we've decided to scrap this PR and try getting to similar end state in a more incremental fashion. Closing this PR to get it off the list while I'm working on that wave of updates.
